File: Object/BoundingVolumeHierarchy.cpp
#include "Object/BoundingVolumeHierarchy.hpp"
#include <cfloat>
#include <algorithm>

namespace Object {
    const Math::Vector splitPlanes[3] = { Math::Vector(1, 0, 0), Math::Vector(0, 1, 0), Math::Vector(0, 0, 1) };

    BoundingVolumeHierarchy::BoundingVolumeHierarchy(std::vector<Node> &&nodes)
        : mNodes(std::move(nodes))
    {
    }

    BoundingVolumeHierarchy::BoundingVolumeHierarchy(const std::vector<Math::Point> &points, const std::function<BoundingVolume(unsigned int)> &func)
    {
        std::vector<unsigned int> indices(points.size());
        for (unsigned int i = 0; i < indices.size(); i++) {
            indices[i] = i;
        }

        std::vector<TreeNode> tree;
        tree.reserve(points.size() * 2);
        buildKdTree(points, tree, indices.begin(), indices.end(), 0);

        mNodes.reserve(points.size() * 2);
        computeBounds(tree, func, 0);
    }

    const std::vector<BoundingVolumeHierarchy::Node> &BoundingVolumeHierarchy::nodes() const
    {
        return mNodes;
    }

    bool BoundingVolumeHierarchy::intersect(const BoundingVolume::RayData &rayData, float &maxDistance, bool closest, const std::function<bool(unsigned int, float&)> &func) const
    {
        struct StackEntry {
            int nodeIndex;
            float minDistance;
        };

        StackEntry stack[64];

        bool ret = false;
        int n = 0;
        stack[n].nodeIndex = 0;
        stack[n].minDistance = 0;
        n++;
        do {
            n--;
            int nodeIndex = stack[n].nodeIndex;
            const Node &node = mNodes[nodeIndex];
            float nodeMinimum = stack[n].minDistance;
            
            if(nodeMinimum > maxDistance) {
                continue;
            }

            if (node.index <= 0) {
                int index = -node.index;
                if (func(index, maxDistance)) {
                    ret = true;
                    if(!closest) {
                        break;
                    }
                }
            }
            else {
                unsigned int indices[2] = { nodeIndex + 1, static_cast<unsigned int>(node.index) };
                float minDistances[2];
                float maxDistances[2];
                for (int i = 0; i < 2; i++) {
                    minDistances[i] = FLT_MAX;
                    maxDistances[i] = -FLT_MAX;
                    mNodes[indices[i]].volume.intersectRay(rayData, minDistances[i], maxDistances[i]);
                }

                for (int i = 0; i < 2; i++) {
                    int j = (minDistances[0] >= minDistances[1]) ? i : 1 - i;
                    if(maxDistances[j] > 0) {
                        stack[n].nodeIndex = indices[j];
                        stack[n].minDistance = minDistances[j];
                        n++;
                    }
                }
            }
        } while(n > 0);

        return ret;
    }

    unsigned int BoundingVolumeHierarchy::buildKdTree(const std::vector<Math::Point> &centroids, std::vector<TreeNode> &tree, std::vector<unsigned int>::iterator indicesBegin, std::vector<unsigned int>::iterator indicesEnd, unsigned int splitIndex) const
    {
        tree.push_back(TreeNode());
        unsigned int nodeIndex = static_cast<unsigned int>(tree.size() - 1);
        TreeNode &node = tree[nodeIndex];

        unsigned int numIndices = static_cast<unsigned int>(indicesEnd - indicesBegin);
        if (numIndices == 1) {
            unsigned int triangleIndex = *indicesBegin;
            node.index = -static_cast<int>(triangleIndex);
        }
        else {
            const Math::Vector &splitPlane = splitPlanes[splitIndex];

            auto cmp = [&](const unsigned int &idx0, const unsigned int &idx1) {
                return Math::Vector(centroids[idx0]) * splitPlane < Math::Vector(centroids[idx1]) * splitPlane;
            };

            std::vector<unsigned int>::iterator split = indicesBegin + (indicesEnd - indicesBegin) / 2;
            std::nth_element(indicesBegin, split, indicesEnd, cmp);

            buildKdTree(centroids, tree, indicesBegin, split, (splitIndex + 1) % 3);
            node.index = static_cast<int>(buildKdTree(centroids, tree, split, indicesEnd, (splitIndex + 1) % 3));
        }

        return nodeIndex;
    }

    unsigned int BoundingVolumeHierarchy::computeBounds(const std::vector<TreeNode> &tree, const std::function<BoundingVolume(unsigned int)> &func, unsigned int treeIndex)
    {
        const TreeNode &treeNode = tree[treeIndex];
        mNodes.push_back(Object::BoundingVolumeHierarchy::Node());
        unsigned int nodeIndex = static_cast<unsigned int>(mNodes.size() - 1);
        Object::BoundingVolumeHierarchy::Node &node = mNodes[nodeIndex];

        if (treeNode.index <= 0) {
            unsigned int index = static_cast<unsigned int>(-treeNode.index);

            node.index = -static_cast<int>(index);
            node.volume = func(index);
        }
        else {
            computeBounds(tree, func, treeIndex + 1);
            node.volume.expand(mNodes[nodeIndex + 1].volume);
            node.index = computeBounds(tree, func, static_cast<unsigned int>(treeNode.index));
            node.volume.expand(mNodes[static_cast<unsigned int>(node.index)].volume);
        }

        return nodeIndex;
    }

    void BoundingVolumeHierarchy::writeProxy(BVHNodeProxy *proxy) const
    {
        for(int i=0; i<mNodes.size(); i++) {
            proxy[i].index = mNodes[i].index;
            mNodes[i].volume.writeProxy(proxy[i].volume);
        }
    }
}

File: src/object/bounding_volume_hierarchy.rs
use crate::geo;
use crate::object;

use geo::Point3;
use geo::Vec3;

use object::BoundingVolume;
use object::RayData;

pub struct BvhNode {
    pub volume: BoundingVolume,
    pub index: i32,
}

struct TreeNode {
    index: i32,
}

pub struct BoundingVolumeHierarchy {
    nodes: Vec<BvhNode>,
}

const NUM_SPLIT_PLANES: usize = 3;
const SPLIT_PLANES: [Vec3; NUM_SPLIT_PLANES] = [
    Vec3::new(1.0, 0.0, 0.0),
    Vec3::new(0.0, 1.0, 0.0),
    Vec3::new(0.0, 0.0, 1.0),
];

fn build_kd_tree(
    centroids: &[Point3],
    tree: &mut Vec<TreeNode>,
    indices: &mut [usize],
    split_index: usize,
) -> usize {
    tree.push(TreeNode { index: 0 });
    let node_index = tree.len() - 1;

    if indices.len() == 1 {
        let point_index = indices[0];
        tree[node_index].index = -(point_index as i32);
    } else {
        let split_plane = SPLIT_PLANES[split_index];

        let split_point = indices.len() / 2;
        indices.select_nth_unstable_by(split_point, |a, b| {
            (Vec3::from(centroids[*a]) * split_plane)
                .total_cmp(&(Vec3::from(centroids[*b]) * split_plane))
        });

        build_kd_tree(
            centroids,
            tree,
            &mut indices[..split_point],
            (split_index + 1) % NUM_SPLIT_PLANES,
        );
        tree[node_index].index = build_kd_tree(
            centroids,
            tree,
            &mut indices[split_point..],
            (split_index + 1) % NUM_SPLIT_PLANES,
        ) as i32;
    }

    return node_index;
}

fn compute_bounds<F>(
    nodes: &mut Vec<BvhNode>,
    tree: &[TreeNode],
    func: &F,
    tree_index: usize,
) -> usize
where
    F: Fn(usize) -> BoundingVolume,
{
    let tree_node = &tree[tree_index];
    nodes.push(BvhNode {
        volume: BoundingVolume::new(),
        index: 0,
    });
    let node_index = nodes.len() - 1;

    if tree_node.index <= 0 {
        let index = -tree_node.index as usize;
        nodes[node_index].index = -(index as i32);
        nodes[node_index].volume = func(index);
    } else {
        compute_bounds(nodes, tree, func, tree_index + 1);
        let volume0 = nodes[node_index + 1].volume;
        nodes[node_index].volume.include_volume(volume0);

        let new_index = compute_bounds(nodes, tree, func, tree_node.index as usize);
        nodes[node_index].index = new_index as i32;

        let volume1 = nodes[new_index].volume;
        nodes[node_index].volume.include_volume(volume1);
    }

    return node_index;
}

impl BoundingVolumeHierarchy {
    pub fn with_volumes<F>(points: &[Point3], func: &F) -> BoundingVolumeHierarchy
    where
        F: Fn(usize) -> BoundingVolume,
    {
        let mut indices = Vec::with_capacity(points.len());
        for i in 0..points.len() {
            indices.push(i);
        }

        let mut tree = Vec::with_capacity(points.len() * 2);
        build_kd_tree(&points, &mut tree, &mut indices[..], 0);

        let mut nodes = Vec::with_capacity(points.len() * 2);
        compute_bounds(&mut nodes, &tree, func, 0);

        return BoundingVolumeHierarchy { nodes };
    }

    pub fn with_nodes(nodes: Vec<BvhNode>) -> BoundingVolumeHierarchy {
        return BoundingVolumeHierarchy { nodes };
    }

    pub fn intersect<F>(&self, raydata: RayData, max_distance: f32, closest: bool, func: &mut F)
    where
        F: FnMut(usize, f32) -> Option<f32>,
    {
        #[derive(Copy, Clone)]
        struct StackEntry {
            node_index: usize,
            min_distance: f32,
        }

        let mut stack = [StackEntry {
            node_index: 0,
            min_distance: 0.0,
        }; 64];

        let mut distance = max_distance;
        let mut n = 0;
        stack[n] = StackEntry {
            node_index: 0,
            min_distance: 0.0,
        };
        n += 1;

        while n > 0 {
            n -= 1;
            let node_index = stack[n].node_index;
            let node = &self.nodes[node_index];
            let node_min = stack[n].min_distance;

            if node_min > distance {
                continue;
            }

            if node.index <= 0 {
                let index = (-node.index) as usize;
                if let Some(new_distance) = func(index, distance) {
                    distance = new_distance;
                    if !closest {
                        break;
                    }
                }
            } else {
                let indices = [node_index + 1, node.index as usize];
                let mut min_distances = [0.0; 2];
                let mut max_distances = [0.0; 2];
                for i in 0..2 {
                    if let Some((min, max)) = self.nodes[indices[i]].volume.intersect_ray(raydata) {
                        min_distances[i] = min;
                        max_distances[i] = max;
                    } else {
                        min_distances[i] = f32::MAX;
                        max_distances[i] = -f32::MAX;
                    }
                }

                for i in 0..2 {
                    let j = if min_distances[0] >= min_distances[1] {
                        i
                    } else {
                        1 - i
                    };
                    if max_distances[j] >= 0.0 {
                        stack[n].node_index = indices[j];
                        stack[n].min_distance = min_distances[j];
                        n += 1;
                    }
                }
            }
        }
    }
}
